### ISSUE TITLE
Golang function latency

### ISSUE THREAD

**Issue by mwuertinger** [iovisor/bpftrace#970]
I'm currently playing around with tracing Go programs and I'm trying to measure the latency of a function invocation like this:
```
uprobe:/tmp/app:"main.myfunc"
{
        @start[tid] = nsecs;
}

uretprobe:/tmp/app:"main.myfunc"
/@start[tid]/
{
        @usecs = hist((nsecs - @start[tid]) / 1000);
        delete(@start[tid]);
}
```
I took this code from one of the [examples](https://github.com/iovisor/bpftrace/blob/master/tools/gethostlatency.bt) and modified it accordingly. You can find my full example code here: https://github.com/mwuertinger/golang-bpftrace-playground
I'm using bpftrace `v0.9.2-102-g254b` compiled from source on `Ubuntu 19.04` with kernel `5.0.0-36-generic`.

While the code appears to work fine, I doubt that it produces correct results since Go is using light weight threads (goroutines) which are multiplexed onto a pool of OS threads. Therefore it could happen that the goroutine running on a thread (`tid`) is changing between the `uprobe` and the `uretprobe`, thus producing invalid measurements.

I'm wondering if there currently is any way to work around this? This should also affect other languages which are using light weight threads (eg. Erlang).

**Comment by danobi** [comment]
Is there anywhere in the go scheduler you can hook into? If you first see a function entry on a tid and then see a context switch but no function exit on the same tid afterwards, could that give you a hint to where your goroutine travelled?

**Comment by fbs** [comment]
I was going to suggest using the stack pointer as each go routine has it's own stack. But that doesn't guard you against stack resizes which copy the stack and modify all pointers (also on the stack).

But uretprobes + golang is not the best idea:

```
package main
import (
  "fmt"
  "time"
)

func fn(t int) {
  fmt.Printf("Data2: %d\n", t)
  time.Sleep(1 * time.Second)
}

func main() {
  go fn(0x555555)
  go fn(0x989898)
  time.Sleep(2*time.Second)
}

$ bpftrace -e 'u:./test:"main.fn" { printf("%s, content %llx, stack %llx, stack-1: %llx\n", probe, sarg0, reg("sp"), *(reg("sp"))); } ur:./test:"main.fn" { }' -c ./test
Attaching 2 probes...
uprobe:./test:main.fn, content 555555, stack c42002ffd0, stack-1: 44edc1
uprobe:./test:main.fn, content 989898, stack c4200307d0, stack-1: 44edc1
Data2: 10000536
Data2: 5592405
SIGILL: illegal instruction
PC=0x7fffffffe001 m=3 sigcode=0

goroutine 6 [running]:
runtime: unknown pc 0x7fffffffe001
stack: frame={sp:0xc4200307d8, fp:0x0} stack=[0xc420030000,0xc420030800)
000000c4200306d8:  000000000052d7a0  00000000004b434e
000000c4200306e8:  0000000000000005  0000000000000013
000000c4200306f8:  0000000000000002  000000c420030708
000000c420030708:  00000000004407a6 <time.Sleep+358>  000000000052d7a0
...
runtime: unknown pc 0x7fffffffe001
```

Which is slightly sub optimal (caused by uretprobes modifying the stack)

https://github.com/iovisor/bcc/issues/1320#issuecomment-407927542 has some useful info regarding go tracing. 

Tl;dr is that you need:

1. A unique ID to trace your go routines on
2. Uprobes on all `ret` instructions  instead of a single uretprobe

For 1. haven't looked into that yet but https://blog.sgmansfield.com/2015/12/goroutine-ids/ might work
For 2., we just landed uprobe offset support. I noticed that it doesn't support `u:./test:"main.fn"+172` yet (parser bug), `u:./test:main.fn+172` isn't valid either, I will put up a PR for that.

**Comment by mwuertinger** [comment]
Thanks for the hints, @fbs. So far I haven't found a way to directly access the goroutine ID. There is a function [`getg()`](https://github.com/golang/go/blob/39a9cb4b5dbf1e518b0c66fa3a7b4175f90226fc/src/runtime/stubs.go#L18) but it's replaced by some assembler in the compiler and I didn't yet have time to decipher this.

What I did get to work, however, is to get called back whenever a new goroutine is scheduled:
```
u:./app:"runtime.execute" {
        $goid = *(int64*)(sarg0+19*8);
        printf("[%d] (%d) goroutine %d\n", pid, tid, $goid)
}
```

This could be used to discard measurements for a `tid` when the goroutine is preempted.

**Comment by felixge** [comment]
@mwuertinger nice! For anybody who is looking for a less (but still) hacky approach, you can also define go's internal `g` struct in your bpftrace script so you can access its fields by name:

https://gist.github.com/felixge/7b21f8c3fc3add7a2a3e52dbac65cc53#file-trace-bt

I created my definitions by looking at the runtime source code, but you might also be able to automate it by using `pahole -q -C runtime.g <binary>` or similar.
